### PR TITLE
[SPARK-30245][SQL][FOLLOWUP] Improve regex expression when pattern not changed

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
This PR follows up https://github.com/apache/spark/pull/26875.

### Why are the changes needed?
When pattern is not static, we should avoid compile pattern every time if some pattern is same.

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
Exists UT.

### CODE REVIEW

**Comment by beliefer** [comment]
cc @maropu @dongjoon-hyun @cloud-fan

**Comment by maropu** [comment]
If you're interested in this, could you answer the @HyukjinKwon comment? https://github.com/apache/spark/pull/26875#issuecomment-580062840
I think we should check that performance first.

**Comment by beliefer** [comment]
> If you're interested in this, could you answer the @HyukjinKwon comment? [#26875 (comment)](https://github.com/apache/spark/pull/26875#issuecomment-580062840)
> I think we should check that performance first.

Oh, yeah. I see that. We need more investigation.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala [hunk 1]
```diff
@@ -54,9 +54,20 @@ trait StringRegexExpression extends Expression
...
 
+  var lastPatternStr: String = null
+  var compiledPattern: Pattern = null
+
   def nullSafeMatch(input1: Any, input2: Any): Any = {
-    val s = input2.asInstanceOf[UTF8String].toString
-    val regex = if (cache == null) compile(s) else cache
+    val patternStr = input2.asInstanceOf[UTF8String].toString
+    val regex = if (cache == null) {
+      if (!(patternStr).equals(lastPatternStr)) {
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala]: Benefits of this optimization depend on how patterns are distributed in a column. For example, if we have:

| string | pattern|
|-|-|
|s1|pattern1|
|s2|pattern2|
|s3|pattern1|
|s4|pattern2|

in that case, this optimization can lead to slowdown.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala [hunk 2]
```diff
@@ -54,9 +54,20 @@ trait StringRegexExpression extends Expression
     Pattern.compile(escape(str))
   }
 
+  var lastPatternStr: String = null
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala]: You can keep `lastPattern` as `Pattern`, and compare it to the current pattern via `lastPattern.pattern() == patternStr`. No need to keep `lastPattern` as a string.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala (hunk 1 above)
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala]: Are there any reasons to not use `patternStr != lastPatternStr`?

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala]: oh, I see this is a copy-paste from Java codegen

**Comment by dongjoon-hyun** [comment]
I agree with @maropu and @HyukjinKwon 's opinion. Without resolving their concerns, we should not merge this because this might not be the final follow-up.

**Comment by beliefer** [comment]
@dongjoon-hyun @HyukjinKwon Yes. maybe we need more performance investigation, but I find `RegExpExtract` and `RegExpReplace` also use this method.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala (hunk 2 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala]: Of course，I can do like this. I just followed the way of `RegExpExtract` and `RegExpReplace` here.

**Comment by maropu** [comment]
I will close this and the jira because of https://github.com/apache/spark/pull/26875#issuecomment-601999002

**Comment by beliefer** [comment]
@maropu Thanks!

**Comment by HyukjinKwon** [comment]
Thank you so much for investigations @beliefer.

**Comment by beliefer** [comment]
@HyukjinKwon With pleasure do it.

**Pull request closed without merging**